### PR TITLE
[Account Settings] Add Ability to Change Username

### PR DESCRIPTION
**Pull request by danielebogo**
Refs. #9705

username-change-2

This PR adds the ability to change the username. The feature is based on the _Flux + View Model_ pattern. The view controller is a subclass of `SignupUsernameTableViewController` in order to keep the same UI structure.

## To test:
- Select the Tab _Me_ -> _Account Settings_
- Tap on the _username_ to open the new screen and change the username
- If the user is an A8C user you shouldn't be able allowed to change the username

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Me/Change Username/ChangeUsernameViewController.swift [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        usernameTextfield.resignFirstResponder()
+    }
+}
+
+private extension ChangeUsernameViewController {
+    func setupViewModel() {
+        viewModel.reachabilityListener = { [weak self] in
+            DispatchQueue.main.async {
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Me/Change Username/ChangeUsernameViewController.swift]: Should we just ensure that each of these listener blocks is called on the main queue from the viewModel?

File: WordPress/Classes/Stores/AccountSettingsStore.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
+import WordPressFlux
+
+enum AccountSettingsState: Equatable {
+    case stationary
```
[the remark this answers is not in the document]
Comment (danielebogo) [reply WordPress/Classes/Stores/AccountSettingsStore.swift]: Oh I like `idle`!!! Thanks a lot!

**Comment by danielebogo** [comment]
Thanks @frosty for reviewing it. I update everything. I also added the local validation considering the alphanumeric charset.

I added the alert to confirm the save action. I'm not sure about the copy though.

CC @SylvesterWilmott 

Simulator Screen Shot - iPhone X - 2019-08-01 at 16 25 16

Simulator Screen Shot - iPhone X - 2019-08-01 at 17 20 29

File: WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift [hunk 3]
```diff
@@ -132,6 +143,13 @@ private extension ChangeUsernameViewModel {
...
             static let saveUsername = NSLocalizedString("There was an error saving the username", comment: "Text displayed when there is a failure saving the username.")
             static let emptyValue = NSLocalizedString("Username must be at least 4 characters.", comment: "Error displayed if the username input is an empty string")
+            static let alphanumeric = NSLocalizedString("Username can only contain lowercase letters (a-z) and numbers.", comment: "Error displayed if the username doesn't contain numbers or digits")
         }
     }
 }
+
+private extension String {
+    var isAlphanumeric: Bool {
+        return range(of: "[^a-zA-Z0-9]", options: .regularExpression) == nil
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift]: Are usernames allowed accented characters at all? (I don't know the answer to this) I suspect not, but if they are we may want to do something like

`rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) == nil`

Comment (danielebogo) [reply WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift]: >rangeOfCharacter(from: CharacterSet.alphanumerics.inverted) == nil

This was my first attempt. Using characters sets to check the string, but it includes the accents. The regex then was the simple solution to get only alphanumerics characters ignoring diacritics.

**Comment by danielebogo** [comment]
@frosty @SylvesterWilmott Calypso uses to add a confirmation textfield as soon you start typing in the _Username_ field. Although I think this is still the best experience, because with an alert the user can accidentally confirm the action, a second textfield (plus the actions row we can add in future) pushes the text to the bottom below the keyboard.
I was thinking if at this point we can use a button at the bottom of the page rather than the save button in the navigation bar. This forces the user to scroll the page and read the paragraph.
What do you think?

**Comment by folletto** [comment]
Any reason why we aren't reusing the name suggestions from the signup flow to avoid people getting frustrated in guessing an available username?

**Comment by danielebogo** [comment]
@folletto  we we are using the same experience of Calypso. There's an API call to validate what you type but not to suggest a new username. Should we stay with it or make a different flow? @SylvesterWilmott @frosty

**Comment by folletto** [comment]
I haven't checked in a while so I'm not sure, but we already have a screen to do that in the signup flow! Already designed, with suggestions. 

See #8711 

To me _this_ is introducing a different flow, not the other way around, as the users are on mobile, and have already (maybe) experienced the other UI.

File: WordPress/Classes/ViewRelated/Me/Change Username/ChangeUsernameViewController.swift [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    enum Constants {
+        static let actionButtonTitle = NSLocalizedString("Save", comment: "Settings Text save button title")
+        static let username = NSLocalizedString("Username", comment: "The header and main title")
+        static let cellIdentifier = "SearchTableViewCell"
+
+        enum Alert {
+            static let loading = NSLocalizedString("Loading usernames", comment: "Shown while the app waits for the username suggestions web service to return during the site creation process.")
+            static let title = NSLocalizedString("Careful!", comment: "Alert title.")
+            static let message = NSLocalizedString("You are changing your Username to %@. Changing your username will also affect your Gravatar profile and IntenseDebate profile addresses. \nConfirm your new Username to continue.", comment: "Alert message.")
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Me/Change Username/ChangeUsernameViewController.swift]: Nitpick, but a couple of occurrences of 'username' in here have a capital U. I think they should all be lowercase?

File: WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,140 @@
+import Reachability
+import WordPressFlux
+
+class ChangeUsernameViewModel {
+    typealias VoidListener = () -> Void
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Me/Change Username/View Model/ChangeUsernameViewModel.swift]: Perhaps just `Listener`?

File: WordPress/Classes/Services/AccountSettingsService.swift [hunk 6]
```diff
@@ -18,6 +18,7 @@ protocol AccountSettingsRemoteInterface {
     func getSettings(success: @escaping (AccountSettings) -> Void, failure: @escaping (Error) -> Void)
     func updateSetting(_ change: AccountSettingsChange, success: @escaping () -> Void, failure: @escaping (Error) -> Void)
     func changeUsername(to username: String, success: @escaping () -> Void, failure: @escaping () -> Void)
+    func validateUsername(to username: String, success: @escaping () -> Void, failure: @escaping (Error) -> Void)
```
Comment (frosty) [inline WordPress/Classes/Services/AccountSettingsService.swift]: Are we still using the validate methods now we changed to the suggestion model? If not, should we remove them?

**Comment by frosty** [comment]
> Should we disable the Save button until a different (new) username is selected?

I can't recreate this at the moment, so I'm not quite sure how I got into that state 🤔

One other thing I forgot to mention: on the confirmation alert, if you type the new username incorrectly we just dismiss the alert and fail silently. We should probably let the user know that their username wasn't changed.

**Comment by danielebogo** [comment]
Hey @frosty thanks for your review! I pushed some change and it's ready for another round!
Here you can find some screenshot about the alert with the button disabled/enabled:

Simulator Screen Shot - iPhone X - 2019-08-08 at 15 47 02

Simulator Screen Shot - iPhone X - 2019-08-08 at 15 47 06

**Comment by danielebogo** [comment]
Thanks @frosty ! It's ready again 😊

**Pull request merged**